### ISSUE TITLE
Repo object has empty list of members in post_setup

### ISSUE THREAD

**Issue by htdebeer** [repobee/repobee#769]
I have written a `post_setup` hook. When it is run, the repo object is as follows:

```bash
StudentRepo(name='someid-test1', team=StudentTeam(members=[], name=''), url='https://gitlab.tue.nl/git-for-assignments-course2020/someid/someid-test1.git', _path=None)
```

The repository has been created correctly, so I would expect `someid` to show up as the sole element in `members`, but `members` is empty.

I call repobee as follows:

```bash
repobee -p canvas -p gitlab repos setup  --canvas-assignment-id 54427 --assignments test1 --students-file single_student.lst --traceback
```

The file `single_student.lst` contains a single line with `someid` on it. 

When I run the `repos clone` command, the `members` field is filled with `someid` as expected.

It is unclear what is going wrong because the `clone` hook is working as expected. Do you have any idea?

**Comment by slarse** [comment]
Does the actual team (on GitLab) get the member? Most likely, the local student team isn't updated internally in RepoBee. Actually, I'm pretty sure it isn't. So there's probably nothing wrong in the setting up of the teams, just the local state. I'll have a look at this over the weekend.

**Comment by htdebeer** [comment]
> Does the actual team (on GitLab) get the member? 

No! Actually, the only member assigned to the team is me. None of the students are getting assigned. So, that's something I can investigate on my end. It is possible these students do not have an activated account on the gitlab instance, although they are able to login to it with their school's account.

If that's the issue, this would mean we have to do some extra administration on our end before we can use Repobee.

**Comment by slarse** [comment]
> It is possible these students do not have an activated account on the gitlab instance

That would definitely cause this, as RepoBee cannot create accounts. It can just assign existing ones. There still might be a local update issue as well, though, so I'll look into it anyway. I haven't used the `post_setup` hook much so it may not be super well polished.

I'm not sure how your school handles account creation, but at KTH students get their accounts created the first time they log in. Typically, we re-execute the setup command once every few days in the beginning for stragglers to get access to their repos.

**Comment by htdebeer** [comment]
Yes, I believe our systems behave the same as yours. 

However, whenever I try to `setup` the same repo twice, I get the error:

```
[ERROR] 400: The project sharing with group is disabled
[ERROR] Critical exception
```

caused by:

```
gitlab.exceptions.GitlabHttpError: 400: {'name': ['has already been taken'], 'path': ['has already been taken'], 'limit_reached': []}
``

This looks like some configuration error on my part, but I am not sure what it is. In the gitlab docs I find this [article](https://docs.gitlab.com/ee/user/project/members/share_project_with_groups.html), but that does not seem related.

**Comment by slarse** [comment]
Hmm, that looks like a configuration thing on the actual GitLab platform. Possibly in the course group. What happens if you try to share a project with a group manually, through the web interface? Is that possible?

**Comment by slarse** [comment]
Oh, and that article is related. RepoBee student groups access to their repos by sharing the project.

**Comment by htdebeer** [comment]
Yes, apparently the whole sharing of groups was disabled, and as a result there was no sign of sharing groups in a project's interface.

I disabled the option

> Prevent sharing a project within git-for-assignments-course2020 with other groups This setting will be applied to all subgroups unless overridden by a group owner. Groups that already have access to the project will continue to have access unless removed manually.

on the course's main group and then I deleted all already created groups and projects. Now I can rerun the `setup` command as often as I like!

However, the first time I run the command, the repositories are created and I get the error that there are no students in my `post_setup` method. The second and third time I do not get any such message even though the students do still not exist on our gitlab instance. 

Does this mean that the `post_setup` hook is not run the second and third time because no repository is being created because it is already created during the first time?

**Comment by slarse** [comment]
Oh wow you responded seconds before I was done with my lengthy message :D. Shorter version: the empty student teams are caused by students not existing. The members list being empty is expected, the name of the team being the empty string I consider a bug to be fixed.

I also came to the same conclusion as you did with the "disable sharing", that's gonna have to go into the docs, or put a check in (if possible).

> Does this mean that the post_setup hook is not run the second and third time because no repository is being created because it is already created during the first time?

Yes, I have no idea of what I was thinking there. I mean, it's not an oversight, I've explicitly stated this in the docs: https://repobee.readthedocs.io/en/latest/repobee_plug/code.html#repobee_plug._exthooks.post_setup

I must have somehow thought it was a good idea, but that's absolutely terrible design in hindsight. I must update that hook and break the API. I'm thinking something like this:

```python
def post_setup(repo: StudentRepo, newly_created: bool, api: PlatformAPI):
```

where `newly_created` is `True` if the repository was created by this invocation, and `False` if it already existed.

Alternatively, I might split the hook in two, `post_setup_existing`, and `post_setup_new`. What do you think: two different hooks or parameter for the old hook?

**Comment by slarse** [comment]
Adding the `newly_created` parameter seems like the best option to me, even though it's breaking.

**Comment by htdebeer** [comment]
Yes. To limit the breakage, you can move the new argument to the end. That way, existing hooks keep working as is. Your second solution only makes sense when you expect most of your users to want to do separate things in both cases, but I suspect either they do not care about the distinction or, like in my case, they want to try to do the same thing. 

With this new argument I can make the integration with Canvas work! Thanks!

**Comment by slarse** [comment]
> To limit the breakage, you can move the new argument to the end.

I could do that, but I personally prefer having parameters ordered by relevance. In the case of the extension hook functions, the `api` parameter is by far the least important in terms of information gain.

But I suppose that I could make an exception to that in order not to break the API.

**Comment by htdebeer** [comment]
That's a choice you have to make. And whatever you decide, it does set a precedent. Cannot add you the parameter as a field to the StudentRepo field? For example, using an enum (does Python have those?) to differentiate between created and updated? Or make it a boolean `created` field, that is only true the first time round.

Alternatively, you can introduce a new hook method with the extra parameter and mark the current one as "deprecated". Of course, because you named it well, you end up with a less clear name for the new method.

**Comment by slarse** [comment]
> Cannot add you the parameter as a field to the StudentRepo field?

Not really, because `StudentRepo` is used where it does not make sense to say whether or not it was created. For example, in the `post_clone`, that attribute just doesn't make sense.

I think I'll go with the backwards-compatible way of appending arguments, as that's really how pluggy is meant to be used. It does have one design-related upside: all extension hooks that operate on repos have `repo` and `api` as the first two arguments, making any succeeding arguments stick out more.

> Alternatively, you can introduce a new hook method with the extra parameter and mark the current one as "deprecated".

The downside of possibly having strangely ordered arguments seems small compared to having a bunch of deprecated hooks. This is very likely going to happen again, and I think the technical debt of having possibly strangely ordered arguments in the hooks is much less of a problem than having a bunch of deprecated hooks. What do you think?

**Comment by htdebeer** [comment]
I agree with you. You can always do a major API clean-up release after a while or with another big API change anyway.

**Comment by slarse** [comment]
I realized a slight inconvenience. Although passing both fresh and preexisting repos to the hook and adding a parameter to indicate which is which is not overtly breaking, it is instead silently breaking. Repositories that wouldn't be passed previously now are.

The only way to address this in a non-breaking fashion would be to add a new hook, e.g. `post_setup_preexisting` for preexisting repositories. I think that having a single hook is more user-friendly, though, so I'll go with that. All existing implementations of the hook that I know about won't be affected anyway.

**Comment by slarse** [comment]
I'm closing this now as the major points have been addressed, and new issues #771 and #772 document the two remaining issues.
